# Análise de Vendas - Pedidos, Clientes e Reviews
Por: Matheus Mendonça Lopes

Para esse projeto, serão analisadas três das tabelas fornecidas pelo *Kaggle*, que são:<br> 
- `olist_orders_dataset.csv`, contendo os pedidos do *e-commerce*; 
- `olist_customers_dataset.csv`: contendo os clientes da plataforma;
- `olist_order_reviews_dataset.csv`: contendo os reviews de cada pedido;
<br>
Para cada tabela, será realizada uma análise exploratória inicial;
<br>
Em seguida, será verificado:
- Quais são os 20 clientes que mais realizaram pedidos;
- Desses 20 clientes, quais não deixam reviews;
- Se pedidos com avaliações abaixo da média normalmente têm avaliações escritas para a empresa saber onde aprimorar.

## Importando DataFrames

In [17]:
import numpy as np
import pandas as pd

In [18]:
df_pedidos_raw = pd.read_csv('../data/olist_orders_dataset.csv')
df_pedidos_raw.head(10)

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
5  a4591c265e18cb1dcee52889e2d8acc3  503740e9ca751ccdda7ba28e9ab8f608   
6  136cce7faa42fdb2cefd53fdc79a6098  ed0271e0b7da060a393796590e7b737a   
7  6514b8ad8028c9f2cc2374ded245783f  9bdf08b4b3b52b5526ff42d37d47f222   
8  76c6e866289321a7c93b82b54852dc33  f54a9f0e6b351c431402b8461ea51999   
9  e69bfb5eb88e0ed6a785585b27e16dbf  31ad1d1b63eb9962463f764d4e6e0c9d   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   
5    delivered      2017-07-09 21:57:05  2017-07-09 22:10:13   
6     invoiced      2017-04-11 12:22:08  2017-04-13 13:25:17   
7    delivered      2017-05-16 13:10:30  2017-05-16 13:22:11   
8    delivered      2017-01-23 18:29:09  2017-01-25 02:50:47   
9    delivered      2017-07-29 11:55:02  2017-07-29 12:05:32   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   
5          2017-07-11 14:58:04           2017-07-26 10:57:55   
6                          NaN                           NaN   
7          2017-05-22 10:07:46           2017-05-26 12:55:51   
8          2017-01-26 14:16:31           2017-02-02 14:08:10   
9          2017-08-10 19:45:24           2017-08-16 17:14:30   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00  
5           2017-08-01 00:00:00  
6           2017-05-09 00:00:00  
7           2017-06-07 00:00:00  
8           2017-03-06 00:00:00  
9           2017-08-23 00:00:00

In [38]:
df_clientes_raw = pd.read_csv('../data/olist_customers_dataset.csv')
df_clientes_raw.head(10)

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
5  879864dab9bc3047522c92c82e1212b8  4c93744516667ad3b8f1fb645a3116a4   
6  fd826e7cf63160e536e0908c76c3f441  addec96d2e059c80c30fe6871d30d177   
7  5e274e7a0c3809e14aba7ad5aae0d407  57b2a98a409812fe9618067b6b8ebe4f   
8  5adf08e34b2e993982a47070956c5c65  1175e95fb47ddff9de6b2b06188f7e0d   
9  4b7139f34592b3a31687243a302fa75b  9afe194fb833f79e300e37e580171f22   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP  
5                     89254         jaragua do sul             SC  
6                      4534              sao paulo             SP  
7                     35182                timoteo             MG  
8                     81560               curitiba             PR  
9                     30575         belo horizonte             MG

In [39]:
df_reviews_pedidos_raw = pd.read_csv('../data/olist_order_reviews_dataset.csv')
df_reviews_pedidos_raw.head(10)

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   
5  15197aa66ff4d0650b5434f1b46cda19  b18dcdf73be66366873cd26c5724d1dc   
6  07f9bee5d1b850860defd761afa7ff16  e48aa0d2dcec3a2e87348811bcfdf22b   
7  7c6400515c67679fbee952a7525281ef  c31a859e34e3adac22f376954e19b39d   
8  a3f6f7f6f433de0aefbb97da197c554c  9c214ac970e84273583ab523dfafd09b   
9  8670d52e15e00043ae7de4c01cc2fe06  b9bf720beb4ab3728760088589c62129   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   
5             1                  NaN   
6             5                  NaN   
7             5                  NaN   
8             5                  NaN   
9             4            recomendo   

                              review_comment_message review_creation_date  \
0                                                NaN  2018-01-18 00:00:00   
1                                                NaN  2018-03-10 00:00:00   
2                                                NaN  2018-02-17 00:00:00   
3              Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4  Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   
5                                                NaN  2018-04-13 00:00:00   
6                                                NaN  2017-07-16 00:00:00   
7                                                NaN  2018-08-14 00:00:00   
8                                                NaN  2017-05-17 00:00:00   
9  aparelho eficiente. no site a marca do aparelh...  2018-05-22 00:00:00   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53  
5     2018-04-16 00:39:37  
6     2017-07-18 19:30:34  
7     2018-08-14 21:36:06  
8     2017-05-18 12:05:37  
9     2018-05-23 16:45:47

## Análises Iniciais

#### Análise Inicial: Pedidos

In [40]:
df_pedidos_raw.shape

(99441, 8)

Podemos observar que temos uma tabela de 99441 linhas e 8 colunas. Os próximos passos da análise inicial devem ser:
    - Verificar quais são as colunas;
    - Verificar se a coluna de `ids` possui `NaN` ou alguma duplicata;
    - Trabalhar com um novo dataset sem colunas que não pertencem ao escopo.


In [41]:
df_pedidos_raw.columns

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date'],
      dtype='object')

In [42]:
df_pedidos_raw['order_id'].nunique()

99441

In [43]:
df_pedidos_raw['order_id'].isna().sum()

0

Pode-se perceber que não há `NaN` na coluna de `order_id`, e todos seus valores são únicos. Portanto, se trata da chave primária de pedidos.<br>
Agora, as colunas ``order_purchase_timestamp``, ``order_approved_at``, ``order_delivered_carrier_date``, ``order_delivered_customer_date`` e ``order_estimated_delivery_date`` serão eliminadas para facilitar a leitura, considerando que elas não fazem parte do escopo da análise desejada.

In [44]:
df_pedidos = df_pedidos_raw[['order_id', 'customer_id', 'order_status']]
df_pedidos.head(10)

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
5  a4591c265e18cb1dcee52889e2d8acc3  503740e9ca751ccdda7ba28e9ab8f608   
6  136cce7faa42fdb2cefd53fdc79a6098  ed0271e0b7da060a393796590e7b737a   
7  6514b8ad8028c9f2cc2374ded245783f  9bdf08b4b3b52b5526ff42d37d47f222   
8  76c6e866289321a7c93b82b54852dc33  f54a9f0e6b351c431402b8461ea51999   
9  e69bfb5eb88e0ed6a785585b27e16dbf  31ad1d1b63eb9962463f764d4e6e0c9d   

  order_status  
0    delivered  
1    delivered  
2    delivered  
3    delivered  
4    delivered  
5    delivered  
6     invoiced  
7    delivered  
8    delivered  
9    delivered

É possível identificar que `order_status` já possui valores diferentes pelo `.head()` acima. Portanto, será usado o `.value_counts()` para descobrir quantos valores de cada tipo existem.

In [45]:
df_pedidos['order_status'].value_counts()

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

In [46]:
df_pedidos['order_status'].isna().sum()

0

Para finalizar a análise inicial, será averiguado se, em `customer_id`, há valores repetidos e `NaN`.

In [47]:
df_pedidos['customer_id'].nunique()

99441

In [48]:
df_pedidos['customer_id'].isna().sum()

0

Portanto, as informações relevantes obtidas sobre o DataFrame `df_pedidos`, até o momento, foram:
- Núm. de Linhas: 99441
- Núm. de Colunas (sendo utilizadas): 3
- Colunas:
    * `order_id`:
        - 99441 valores únicos
        - Nenhum NaN
        - Chave primária
    * `customer_id`: 
        - 99441 valores únicos
        - Nenhum NaN
    * `order_status`:
        - 8 tipos de valores
        - Nenhum NaN

#### Análise Inicial: Clientes

In [50]:
df_clientes_raw.head(10)

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
5  879864dab9bc3047522c92c82e1212b8  4c93744516667ad3b8f1fb645a3116a4   
6  fd826e7cf63160e536e0908c76c3f441  addec96d2e059c80c30fe6871d30d177   
7  5e274e7a0c3809e14aba7ad5aae0d407  57b2a98a409812fe9618067b6b8ebe4f   
8  5adf08e34b2e993982a47070956c5c65  1175e95fb47ddff9de6b2b06188f7e0d   
9  4b7139f34592b3a31687243a302fa75b  9afe194fb833f79e300e37e580171f22   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP  
5                     89254         jaragua do sul             SC  
6                      4534              sao paulo             SP  
7                     35182                timoteo             MG  
8                     81560               curitiba             PR  
9                     30575         belo horizonte             MG

A localização do cliente não é pertinente para a análise do projeto, então a tabela será filtrada para conter apenas as colunas `customer_id` e `customer_unique_id`.

In [54]:
df_clientes = df_clientes_raw[['customer_id', 'customer_unique_id']]
df_clientes.head(10)

customer_id                customer_unique_id
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066
5  879864dab9bc3047522c92c82e1212b8  4c93744516667ad3b8f1fb645a3116a4
6  fd826e7cf63160e536e0908c76c3f441  addec96d2e059c80c30fe6871d30d177
7  5e274e7a0c3809e14aba7ad5aae0d407  57b2a98a409812fe9618067b6b8ebe4f
8  5adf08e34b2e993982a47070956c5c65  1175e95fb47ddff9de6b2b06188f7e0d
9  4b7139f34592b3a31687243a302fa75b  9afe194fb833f79e300e37e580171f22

In [57]:
df_clientes.shape

(99441, 2)

Analisando a coluna `customer_unique_id`, verificando quantidade de valores únicos e `NaN`.

In [55]:
df_clientes['customer_unique_id'].nunique()

96096

In [58]:
df_clientes['customer_unique_id'].isna().sum()

0

A quantidade de valores únicos em `customer_unique_id` é menor que o número de linhas, e não há `NaN`; portanto, há valores repetidos.

Verificando informações da coluna `customer_id`. De acordo com a documentação dos datasets, `customer_id` se trata de um id único criado para cada pedido na tabela de pedidos, portanto, é uma chave estrangeira. Por conta disso, um cliente que faz mais de um pedido acaba tendo o `customer_unique_id` repetido na tabela de clientes.
<br>
Com essa informação, será possível descobrir quantos pedidos cada cliente realizou.

In [59]:
df_clientes['customer_id'].nunique()

99441

In [60]:
df_clientes['customer_id'].isna().sum()

0

Portanto, as informações relevantes obtidas sobre o DataFrame `df_clientes`, até o momento, foram:
- Núm. de Linhas: 99441
- Núm. de Colunas (sendo utilizadas): 2
- Colunas:
    * `customer_id`: 
        - 99441 valores únicos
        - Nenhum NaN
        - Chave estrangeira proveniente de `df_pedidos`
    * `customer_unique_id`:
        - 96096 valores únicos
        - Nenhum NaN
        - Se trata do `id` pra cada cliente

#### Análise Inicial: Reviews 

In [63]:
df_reviews_pedidos_raw.head(10)

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   
5  15197aa66ff4d0650b5434f1b46cda19  b18dcdf73be66366873cd26c5724d1dc   
6  07f9bee5d1b850860defd761afa7ff16  e48aa0d2dcec3a2e87348811bcfdf22b   
7  7c6400515c67679fbee952a7525281ef  c31a859e34e3adac22f376954e19b39d   
8  a3f6f7f6f433de0aefbb97da197c554c  9c214ac970e84273583ab523dfafd09b   
9  8670d52e15e00043ae7de4c01cc2fe06  b9bf720beb4ab3728760088589c62129   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   
5             1                  NaN   
6             5                  NaN   
7             5                  NaN   
8             5                  NaN   
9             4            recomendo   

                              review_comment_message review_creation_date  \
0                                                NaN  2018-01-18 00:00:00   
1                                                NaN  2018-03-10 00:00:00   
2                                                NaN  2018-02-17 00:00:00   
3              Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4  Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   
5                                                NaN  2018-04-13 00:00:00   
6                                                NaN  2017-07-16 00:00:00   
7                                                NaN  2018-08-14 00:00:00   
8                                                NaN  2017-05-17 00:00:00   
9  aparelho eficiente. no site a marca do aparelh...  2018-05-22 00:00:00   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53  
5     2018-04-16 00:39:37  
6     2017-07-18 19:30:34  
7     2018-08-14 21:36:06  
8     2017-05-18 12:05:37  
9     2018-05-23 16:45:47

Considerando que datas não são importantes para o escopo da análise, as colunas `review_comment_title`, `review_creation_date` e `review_answer_timestamp` não serão utilizadas.

In [104]:
df_reviews = df_reviews_pedidos_raw[['review_id','order_id','review_score','review_comment_message']]
df_reviews.head(10)

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   
5  15197aa66ff4d0650b5434f1b46cda19  b18dcdf73be66366873cd26c5724d1dc   
6  07f9bee5d1b850860defd761afa7ff16  e48aa0d2dcec3a2e87348811bcfdf22b   
7  7c6400515c67679fbee952a7525281ef  c31a859e34e3adac22f376954e19b39d   
8  a3f6f7f6f433de0aefbb97da197c554c  9c214ac970e84273583ab523dfafd09b   
9  8670d52e15e00043ae7de4c01cc2fe06  b9bf720beb4ab3728760088589c62129   

   review_score                             review_comment_message  
0             4                                                NaN  
1             5                                                NaN  
2             5                                                NaN  
3             5              Recebi bem antes do prazo estipulado.  
4             5  Parabéns lojas lannister adorei comprar pela I...  
5             1                                                NaN  
6             5                                                NaN  
7             5                                                NaN  
8             5                                                NaN  
9             4  aparelho eficiente. no site a marca do aparelh...

In [105]:
df_reviews.shape

(99224, 4)

Analisando as colunas `review_id` e `order_id`:

In [106]:
df_reviews['review_id'].nunique()

98410

In [107]:
df_reviews['review_id'].isna().sum()

0

In [108]:
df_reviews['order_id'].nunique()

98673

In [109]:
df_reviews['order_id'].isna().sum()

0

É possível que nenhuma das duas colunas possui `NaN`, mas existem valores em cada uma das colunas que se repete.<br>
Verificando o valor mais comum de cada uma delas e como aparecem na tabela:

In [110]:
most_common_review_id = df_reviews['review_id'].value_counts().nlargest(1).index[0]
most_common_review_id

'7b606b0d57b078384f0b58eac1d41d78'

In [111]:
df_reviews[df_reviews['review_id']==most_common_review_id]

review_id                          order_id  \
7500   7b606b0d57b078384f0b58eac1d41d78  f3028a8f41ea1ee2b461420913663f97   
59859  7b606b0d57b078384f0b58eac1d41d78  2deb17060fc1ce18a85eba953ddcdeaf   
61069  7b606b0d57b078384f0b58eac1d41d78  2f8f31eb2f7b6572836d662a6625c8e4   

       review_score review_comment_message  
7500              5                    NaN  
59859             5                    NaN  
61069             5                    NaN

Uma mesmo review pode ser utilizada pelo cliente para diferentes pedidos.

In [112]:
most_common_order_id = df_reviews['order_id'].value_counts().nlargest(1).index[0]
most_common_order_id

'c88b1d1b157a9999ce368f218a407141'

In [113]:
df_reviews[df_reviews['order_id']==most_common_order_id]

review_id                          order_id  \
1985   ffb8cff872a625632ac983eb1f88843c  c88b1d1b157a9999ce368f218a407141   
82525  202b5f44d09cd3cfc0d6bd12f01b044c  c88b1d1b157a9999ce368f218a407141   
89360  fb96ea2ef8cce1c888f4d45c8e22b793  c88b1d1b157a9999ce368f218a407141   

       review_score review_comment_message  
1985              3                    NaN  
82525             5                    NaN  
89360             5                    NaN

O usuário pode deixar diferentes reviews para um mesmo produto.

Analisando a coluna de `review_score`:

In [114]:
df_reviews['review_score'].value_counts(True)*100

review_score
5    57.776344
4    19.291704
1    11.513344
3     8.242965
2     3.175643
Name: proportion, dtype: float64

In [115]:
df_reviews['review_score'].describe()

count    99224.000000
mean         4.086421
std          1.347579
min          1.000000
25%          4.000000
50%          5.000000
75%          5.000000
max          5.000000
Name: review_score, dtype: float64

Percebe-se que a nota média de avaliações é muito próxima de 4, ainda que mais de 50% das avaliações seja 5.

Verificando a porcentagem de `NaN` na coluna `review_comment_message`.

In [126]:
nan_comment_percentage = 100*df_reviews['review_comment_message'].isna().sum()/99224
print(f"A porcetagem de NaN é: {nan_comment_percentage:.2f}%")

A porcetagem de NaN é: 58.70%
